File: core/src/ru/myitschool/dcrawler/skills/targeting/WaveDisplayer.java
package ru.myitschool.dcrawler.skills.targeting;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.ai.AIUtils;
import ru.myitschool.dcrawler.skills.Skill;
import ru.myitschool.dcrawler.utils.AdvancedArray;

/**
 * Created by Voyager on 12.03.2018.
 */
public class WaveDisplayer implements TargetDisplayer {

    @Override
    public boolean targetCreation(int x, int y, Array<TilePos> array, Skill skill) {
        int range = skill.getRange();
        AdvancedArray<Vector2> poses = AIUtils.getCellRaytrace(skill.getDoer().getTileX(), skill.getDoer().getTileY(), x, y, range - 1);
        poses.clip(poses.size - range, poses.size - 1);
        for (Vector2 pos : poses){
            array.add(new TilePos((int) pos.x, (int) pos.y, TargetRenderer.getDefaultTargetTile()));
        }
        return true;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/targeting/WaveDisplayer.kt
package com.wolg_vlad.dcrawler.entities.skills.targeting

import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.entities.ai.AIUtils
import com.wolg_vlad.dcrawler.entities.skills.Skill

/**
 * Created by Voyager on 12.03.2018.
 */
class WaveDisplayer : TargetDisplayer {
    override fun targetCreation(x: Int, y: Int, array: MutableList<TilePos>, skill: Skill): Boolean {
        val range = skill.range
        val poses = AIUtils.getCellRaytrace(skill.doer.tileX, skill.doer.tileY, x, y, range - 1)
        poses!!.clip(poses.size - range, poses.size - 1)
        for (pos in poses) {
            array.add(TilePos(pos!!.x.toInt(), pos.y.toInt(), TargetRenderer.defaultTargetTile))
        }
        return true
    }
}